src/AzureExtension/DevBox/Helpers/WaitingForUserAdaptiveCardSession.cs

```diff
@@ -70,21 +81,43 @@ private static string ConvertIconToDataString(string fileName)
     public IAsyncOperation<ProviderOperationResult> OnAction(string action, string inputs)
     {
         return Task.Run(() =>
-        {
-            var state = _extensionAdaptiveCard?.State;
-            ProviderOperationResult operationResult;
-            if (state == "WaitingForUserSession")
-            {
-                operationResult = new ProviderOperationResult(ProviderOperationStatus.Success, null, null, null);
-                _resumeEvent.Set();
-                Stopped?.Invoke(this, new(operationResult, string.Empty));
-            }
-            else
-            {
-                operationResult = new ProviderOperationResult(ProviderOperationStatus.Failure, null, "Something went wrong", $"Unexpected state:{_extensionAdaptiveCard?.State}");
-            }
+        {
+            try
+            {
+                var state = _extensionAdaptiveCard?.State;
+                ProviderOperationResult operationResult;
+                var data = JsonSerializer.Deserialize<AdaptiveCardJSONToCSClass>(action, _serializerOptions);
+                if (state == "WaitingForUserSession" && data != null)
+                {
+                    switch (data.Id)
+                    {
+                        case "launchAction":
+                            operationResult = new ProviderOperationResult(ProviderOperationStatus.Success, null, null, null);
+                            _launchEvent.Set();
+                            _resumeEvent.Set();
+                            break;
+                        case "resumeAction":
+                            operationResult = new ProviderOperationResult(ProviderOperationStatus.Success, null, null, null);
+                            _resumeEvent.Set();
+                            break;
+                        default:
+                            operationResult = new ProviderOperationResult(ProviderOperationStatus.Failure, null, "Something went wrong", $"Unexpected action:{data.Id}");
+                            break;
+                    }
 
-            return operationResult;
+                    Stopped?.Invoke(this, new(operationResult, string.Empty));

```

Comment:
Do we need to call it if an exception was thrown? (move it to "finally" statement?)
Also, "?" operator is not needed since Stopped is not nullable (though I am not a fan of initializing event to an empty handler).

Reply:
Probably not. 
Good catch with the '?' operator.
Changed.